### PR TITLE
Issue #2140 - Handle reading null from xx_RESOURCES.DATA

### PR DESCRIPTION
**Pull request by tbieste**
Issue #2140 - Handle reading null from xx_RESOURCES.DATA

Signed-off-by: Troy Biesterfeld <tbieste@us.ibm.com>

### CODE REVIEW

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java [hunk 1]
```diff
@@ -1521,119 +1522,167 @@ private TransactionSynchronizationRegistry getTrxSynchRegistry() throws FHIRPers
         List<ExtractedParameterValue> allParameters = new ArrayList<>();
 
         try {
-            map = SearchUtil.extractParameterValues(fhirResource);
+            if (fhirResource != null) {
```
Comment (lmsurpre) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:1525]: this is a case where I'd probably prefer `if (fhirResource == null) { return allParameters }` to keep the nesting down, although I think thats strictly personal preference

Comment (tbieste) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java]: Updated.

File: fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java [hunk 2]
```diff
@@ -2564,20 +2564,27 @@ private Bundle createSearchBundle(List<Resource> resources, FHIRSearchContext se
             }
 
             for (Resource resource : resources) {
-                if (resource.getId() == null) {
-                    throw new IllegalStateException("Returned resources must have an id.");
+                Bundle.Entry.Builder entryBuilder = Bundle.Entry.builder();
```
Comment (punktilious) [inline fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java:2567]: Are there any unit tests covering this? Getting null data for testing probably requires some dedicated test DAO to forcibly set the data value to null.

File: fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java [hunk 3]
```diff
@@ -2564,20 +2564,27 @@ private Bundle createSearchBundle(List<Resource> resources, FHIRSearchContext se
...
+                if (resource != null) {
+                    if (resource.getId() != null) {
+                        entryBuilder.fullUrl(Uri.of(getRequestBaseUri(type) + "/" + resource.getClass().getSimpleName() + "/" + resource.getId()));
+                    } else {
+                        String msg = "A resource with no id was found.";
+                        issues.add(FHIRUtil.buildOperationOutcomeIssue(IssueSeverity.WARNING, IssueType.NOT_SUPPORTED, msg));
```
Comment (lmsurpre) [inline fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java:2573]: do these warnings make it into our operational logs?  if not, we might want to add a log.warning here too because we should really never have this case

File: fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java [hunk 4]
```diff
@@ -2564,20 +2564,27 @@ private Bundle createSearchBundle(List<Resource> resources, FHIRSearchContext se
...
+                    }
+                    entryBuilder.resource(resource);
+                } else {
+                    String msg = "A resource with no data was found.";
+                    issues.add(FHIRUtil.buildOperationOutcomeIssue(IssueSeverity.WARNING, IssueType.NOT_SUPPORTED, msg));
```
Comment (lmsurpre) [inline fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java:2578]: same as above

File: fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java (hunk 3 above)
Comment (tbieste) [reply fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java:2573]: Added.

File: fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java (hunk 4 above)
Comment (tbieste) [reply fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java:2578]: Added.

**Review by lmsurpre — APPROVED** [review]
LGTM

File: fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java (hunk 2 above)
Comment (tbieste) [reply fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java:2567]: No, not specifically. I could, but wasn't sure it was worth the overhead, etc.

Comment (tbieste) [reply fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java:2567]: I added a testcase to FHIRRestHelperTest.java to exercise this code path.

**Comment by tbieste** [comment]
Comments addressed. Ready to merge.

**Review by punktilious — APPROVED** [review]
LGTM
